docs/core/development/policy/compatibility-policy.rst

```diff
@@ -162,24 +162,38 @@ The most basic thing that can happen between Twisted versions, of course, is tha
 That means that no application may ever rely on, for example, the value of any **func_code** object's **co_code** attribute remaining stable, or the **checksum** of a .py file remaining stable.
 
 **Docstrings** may also change at any time.
-No application code may expect any Twisted class, module, or method's __doc__ attribute to remain the same.
+Applications must not depend on any Twisted class, module, or method's metadata attributes such as ``__module__``, ``__name__``, ``__qualname__``, ``__annotations__`` and ``__doc__`` to remain the same.
 
 
 New Attributes
 ^^^^^^^^^^^^^^
 
 New code may also be added.
-No application may ever rely on the output of the ``dir()`` function on any object remaining stable, nor on any object's ``__all__`` attribute, nor on any object's ``__dict__`` not having new keys added to it.
+Applications must not depend on the output of the ``dir()`` function on any object remaining stable, nor on any object's ``__all__`` attribute, nor on any object's ``__dict__`` not having new keys added to it.
 These may happen in any maintenance or bugfix release, no matter how minor.
 
 
 Pickling
 ^^^^^^^^
 
 Even though Python objects can be pickled and unpickled without explicit support for this, whether a particular pickled object can be unpickled after any particular change to the implementation of that object is less certain.
-Because of this, no application may depend on any object defined by Twisted to provide pickle compatibility between any release unless the object explicitly documents this as a feature it has.
+Because of this, applications must not depend on any object defined by Twisted to provide pickle compatibility between any release unless the object explicitly documents this as a feature it has.
 
 
+Representations
+^^^^^^^^^^^^^^^
+
+The printable representations of objects, as returned by ``repr(obj, /)`` and defined by ``def __repr__(self):`` are for debugging and informational purposes.
+Because of this, applications must not depend on any object defined by Twisted to provide repr compatibility between any release.
+descriptors, slots and weakrefs
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+How an object's attributes are defined and accessed is considered an implementation detail.
+To allow backwards compatibility an attribute may be moved from the instance ``__dict__`` into
+an ``@property`` or other descriptor based accessor.
+
+Assigning attributes to a constructed object, or monkey patching, is not considered a public use. This is to allow creating frozen and/or slotted classes. However if an object is weak referencable, an application may depend on it remaining weak referencable
```

Comment:
There are two types of assigning though: you can add new attributes or update existing attributes. For converting to slotted classes, only adding new attributes has to be forbidden.

While I like immutable objects in general, I don't think Twisted's design is prepared for that. I've had to update attributes after construction in multiple cases. While some of those might have been from improper API use or bad code structure, I don't think that covers all of them.

Reply:
> I've had to update attributes after construction in multiple cases. 

update is fine, but add new ones that Twisted never defined?

Reply:
going 'round freezing objects willy nilly seems a bit dangerous. But making everything slotted seems safe. Adding an attribute to an object that twisted is unaware of cannot be a public use of that object

Reply:
I think we agree, but the text itself needs updating:
- applications adding new attributes is not supported, so Twisted can move to slotted classes
- freezing attributes should be handled on a case-by-case basis and will likely require deprecation of a setter property in many cases

Reply:
Careful.  There are cases, I think, where people do depend on being able to attach new attributes on instances of Twisted classes.

In Klein, for example, I have seen (and written) code that attributes a user object to the passed-in `request` object (which comes from `twisted.web`), so you can hold onto something through a request cycle.  Whether that was meant to be allowed or not, breaking this now will create some non-trivial work (perhaps re-architecting stuff) for clients.

Reply:
FWIW `mypy` certainly discourages use of YOLO attributes (I think by default), so once we have everything type hinted, that would be a tool one could use to at least identify some of these use cases upfront.

Reply:
@wsanchez I think twisted.web.request is ripe for being an @attrs.frozen class :D

Reply:
@wsanchez klein seems to use request.setComponent, which is explicit support for runtime instance extension

Reply:
`twisted.web.request` is terrible. Try creating an `IRequest` and then describe how much fun that was. My goal in life is for Klein to stop using it but I've been thwarted by year-long review cycles and then I found type hints and other nonsense to gain more traction.